File: php/tests/canon.test.php
<?php
// JCS canonicalization tests. Vectors mirror the shared corpus
// (spec/corpus/canon.json) and security.md §11.1; byte-for-byte agreement with
// the other implementations over that corpus is checked by bin/canon_check.php.
// Escape-heavy expected strings are built from chr() so the source cannot
// smuggle a raw control byte in place of an escape sequence.
use Bonemesh\Canon;

test('basic sorted keys', function () {
    $cert = ['v' => 3, 'mesh' => 'acme-prod', 'label' => 'alpha', 'idk' => 'YWJj', 'nbf' => 1788500000, 'exp' => 1790000000];
    assertEq('{"exp":1790000000,"idk":"YWJj","label":"alpha","mesh":"acme-prod","nbf":1788500000,"v":3}', Canon::canonicalize($cert));
});

test('sig field is stripped', function () {
    $cert = ['v' => 3, 'mesh' => 'm', 'label' => 'alpha', 'idk' => 'AA==', 'nbf' => 0, 'exp' => 1, 'sig' => 'IGNORED'];
    assertEq('{"exp":1,"idk":"AA==","label":"alpha","mesh":"m","nbf":0,"v":3}', Canon::canonicalize($cert));
});

test('non-ascii emitted as raw utf-8', function () {
    $cert = ['v' => 3, 'mesh' => 'm', 'label' => 'café', 'idk' => 'AA==', 'nbf' => 0, 'exp' => 1];
    assertEq('{"exp":1,"idk":"AA==","label":"café","mesh":"m","nbf":0,"v":3}', Canon::canonicalize($cert));
});

test('quote and backslash escaping', function () {
    $bs = chr(0x5c);
    $cert = ['v' => 3, 'mesh' => 'm', 'label' => 'a"b' . $bs . 'c', 'idk' => 'AA==', 'nbf' => 0, 'exp' => 1];
    $labelTok = '"a' . $bs . '"b' . $bs . $bs . 'c"';
    assertEq('{"exp":1,"idk":"AA==","label":' . $labelTok . ',"mesh":"m","nbf":0,"v":3}', Canon::canonicalize($cert));
});

test('control characters use short and unicode escapes', function () {
    $bs = chr(0x5c);
    $label = 'a' . chr(0x08) . chr(0x09) . chr(0x0a) . chr(0x0c) . chr(0x0d) . chr(0x01) . 'b';
    $labelTok = '"a' . $bs . 'b' . $bs . 't' . $bs . 'n' . $bs . 'f' . $bs . 'r' . $bs . 'u0001b"';
    assertEq('{"label":' . $labelTok . '}', Canon::canonicalize(['label' => $label]));
});

test('negative integer is rejected', function () {
    assertThrows(fn () => Canon::canonicalize(['nbf' => -1]));
});

test('float is rejected', function () {
    assertThrows(fn () => Canon::canonicalize(['v' => 3.5]));
});

test('boolean is rejected', function () {
    assertThrows(fn () => Canon::canonicalize(['ok' => true]));
});

test('json array value is rejected', function () {
    assertThrows(fn () => Canon::canonicalize(['xs' => [1, 2, 3]]));
});

test('nested objects sorted independently', function () {
    assertEq('{"a":{"b":3,"y":2},"z":1}', Canon::canonicalize(['z' => 1, 'a' => ['y' => 2, 'b' => 3]]));
});

File: rust/tests/canon_test.rs
// JCS canonicalization tests. Vectors mirror the shared corpus
// (spec/corpus/canon.json) and security.md §11.1; byte-for-byte agreement with
// the Java, Go, and Elixir canonicalizers over that corpus is checked by
// interop/check-canon-rust.sh.

use bonemesh::canon::canonicalize;
use serde_json::json;

#[test]
fn basic_sorted_keys() {
    let cert = json!({"v":3,"mesh":"acme-prod","label":"alpha","idk":"YWJj","nbf":1_788_500_000_u64,"exp":1_790_000_000_u64});
    assert_eq!(
        canonicalize(&cert).unwrap(),
        r#"{"exp":1790000000,"idk":"YWJj","label":"alpha","mesh":"acme-prod","nbf":1788500000,"v":3}"#
    );
}

#[test]
fn sig_field_is_stripped() {
    let cert = json!({"v":3,"mesh":"m","label":"alpha","idk":"AA==","nbf":0,"exp":1,"sig":"IGNORED"});
    assert_eq!(
        canonicalize(&cert).unwrap(),
        r#"{"exp":1,"idk":"AA==","label":"alpha","mesh":"m","nbf":0,"v":3}"#
    );
}

#[test]
fn non_ascii_emitted_raw_utf8() {
    let cert = json!({"v":3,"mesh":"m","label":"café","idk":"AA==","nbf":0,"exp":1});
    assert_eq!(
        canonicalize(&cert).unwrap(),
        "{\"exp\":1,\"idk\":\"AA==\",\"label\":\"café\",\"mesh\":\"m\",\"nbf\":0,\"v\":3}"
    );
}

#[test]
fn string_escaping_quote_backslash() {
    let cert = json!({"v":3,"mesh":"m","label":"a\"b\\c","idk":"AA==","nbf":0,"exp":1});
    assert_eq!(
        canonicalize(&cert).unwrap(),
        r#"{"exp":1,"idk":"AA==","label":"a\"b\\c","mesh":"m","nbf":0,"v":3}"#
    );
}

#[test]
fn control_chars_short_and_u_forms() {
    // Label bytes: 'a', U+0001, TAB, 'b'. Expected escapes built from bytes so
    // no source escape sequence is relied on.
    let label = String::from_utf8(vec![b'a', 0x01, 0x09, b'b']).unwrap();
    let cert = json!({"v":3,"mesh":"m","label":label,"idk":"AA==","nbf":0,"exp":1});
    let u0001 = String::from_utf8(vec![0x5c, 0x75, 0x30, 0x30, 0x30, 0x31]).unwrap();
    let tab = String::from_utf8(vec![0x5c, 0x74]).unwrap();
    let want = format!(
        r#"{{"exp":1,"idk":"AA==","label":"a{u0001}{tab}b","mesh":"m","nbf":0,"v":3}}"#
    );
    assert_eq!(canonicalize(&cert).unwrap(), want);
}

#[test]
fn negative_integer_is_rejected() {
    let cert = json!({"v":3,"nbf":-1});
    assert!(canonicalize(&cert).is_err());
}
